src/modules/mc_pos_control/MulticopterPositionControl.cpp

```diff
@@ -242,19 +192,37 @@ void MulticopterPositionControl::Run()
 
 		// check if any task is active
 		if (_trajectory_setpoint_sub.update(&setpoint)) {
-			_control.setInputSetpoint(setpoint);
 
-			// check if all local states are valid and map accordingly
-			set_vehicle_states(setpoint.vz);
-			_control.setState(_states);
+			PositionControlStates states {
+				.position = Vector3f{local_pos.x, local_pos.y, local_pos.z},
+				.velocity = velocity,
+				.acceleration = acceleration,
+				.yaw = local_pos.heading,
```

Comment:
Are the components which are not valid for sure set to NAN when they come from the estimator? Such that you can remove:
```
// only set position states if valid and finite	
if (PX4_ISFINITE(_local_pos.x) && PX4_ISFINITE(_local_pos.y) && _local_pos.xy_valid) {	
		_states.position(0) = _local_pos.x;	
	_states.position(1) = _local_pos.y;	
} else {	
	_states.position(0) = _states.position(1) = NAN;	
}
```
This is important information for the controller to determine the feasibility of a particular setpoint combination.

Reply:
This, the acceleration state and the vertical position derivative was the whole point of the separate `PositionControlStates` struct which we can probably purge at some point when we took care of all the points.